### PR TITLE
File refactor

### PR DESCRIPTION
**Pull request by R4PaSs**
I've been talking about this PR for a long time, but I think it's finally ready !

So, a big refactor of the I/O block is in order this time (care not for the diff, it's really not this bad, a lot comes from test outputs).

Basically some classes are now gone (BufferedFileReader, FilterStream and such), and the new hierarchy is layered to lower the maintenance and implementation costs of stream subclasses.
On top of that, Codecs are properly supported and functions like `read_char` now work as intended on code points (which is a nice thing) !

The hierarchy is a bit similar to before as there's Readers and Writers and Duplexes, but there's two levels on which each class works: Byte and Char.

The Byte level is the lowest one, in which each byte is directly taken from the underlying data stream (file, socket, etc.).
Char comes on top of it and does proper character handling based on the Codec assigned to it (which defaults to UTF-8).
On the Codec side, right now only UTF-8 and ISO-8851 are implemented, one because we need it, and the other one which was a bit of a test to see if the current approach was viable and not too hard to implement.

Files are essentially unchanged client-side and the API remains equivalent to what it was before.
A few have however changed:
- StringReader/StringWriter: They are now known as MemoryReader and MemoryWriter, they retain the same semantics as before, but are now working on any type of data (not just Chars)
- Protocol: A new class which is basically a base decorator Stream. Since protocols are often layered, the decorator approach seems best-suited here and are composable as-is.
- Socket/Websocket: Both these streams used to be quite broken before, especially compared to the C-API, our sockets used to block indefinitely on a read, now they have the same behaviour as the C-API (block once until a first read, fill the buffer as much as possible and return when either the buffer is filled or when no more reads can be done without blocking).

One of the key points of this refactor was to ease the maintenance and extensibility of the I/O hierarchy, a new Stream now only has a few key-functions to implement to work properly:
- `read_byte` : Reads one byte from the source
- `read_bytes`: Reads multiple bytes using the same API as sockets (block once until first read and fill as much as possible before returning)

On the Writer-side, the `write` equivalents are to be implemented too.

Char reads and writes are handled by the superclass.

This should simplify our hierarchy, make it work better with Text and encodings and should not impact performance (or at least minimally).

So, performance-wise !

`nitc` does not seem to run slower when we take into account the time as measured bu GNU time, however Valgrind seems to be around 800-900MIr slower.
Most of it is lost on the write_chars since the best way to handle Codecs correctly for now is through the use of ephemeral Bytes instances, which take their toll on the resource-side (see #839).
Some fine-tuning for more performance can be done later to make sure no real losses occur due to files and codecs, though as-is it seems to be performing well enough for now.

Bonus: One example of stream usage with sockets/websockets through a protocol-agnostic chat application, it should work out-of-the-box with both the socket client and any websocket implementation.

### CODE REVIEW

**Comment by privat** [comment]
impressive amount of work.

File: lib/core/codecs/codec_base.nit [hunk 1]
```diff
@@ -24,28 +24,44 @@ module codec_base
...
-end
+	# Adds a string `s` coded as the supported encoding to `b`
+	#
+	# Returns the number of bytes written to `s`
+	fun add_string_to(s: Text, b: Bytes): Int is abstract
+
+	# Size of a codet for the target encoding
+	fun codet_size: Int is abstract
+
+	# How many lookaheads might be required to decode a single char ?
```
Comment (privat) [inline lib/core/codecs/codec_base.nit]: what is the unit of lookaheads? bytes or codet_size?

File: examples/klepet/klepet_server.nit [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+# clients, from diverse communication methods.
+#
+# Right now is supported:
+# * Socket
+# * Websocket
+#
+# The program is also a good way to test any kind of new communication stream.
+# When a new stream kind is supported, you should add it to Klepet and try if
+# some communications are processed correctly.
+module klepet_server
```
Comment (privat) [inline examples/klepet/klepet_server.nit]: Why Slovenian?

File: examples/klepet/klepet_server.nit [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+import socket
+import websocket
+
+var ws_server = new TCPServer(8088)
+var server = new TCPServer(4444)
+server.blocking = false
+ws_server.blocking = false
+
+print "Chat socket server listening on port 4444"
```
Comment (privat) [inline examples/klepet/klepet_server.nit]: "port {server.port}"

File: examples/klepet/klepet_server.nit [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+print "Chat websocket server listening on port 8088"
+
+server.listen(5)
+ws_server.listen(5)
+
+var clients = new HashSet[Duplex]
+var marked_for_removal = new HashSet[Duplex]
+
+loop
+	0.05.sleep
```
Comment (privat) [inline examples/klepet/klepet_server.nit]: why sleep?

**Comment by privat** [comment]
This is nice but I need another look to fully understand the new hierarchies.

File: lib/core/protocol.nit [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+import stream
+
+# A Protocol is supposed to handle a connection through any kind of
+# underlying medium and should comply to a set of rules defined by a
+# formal (or not) specification.
+class Protocol
+	super Stream
+	# Which class should a protocol wrap ?
+	type PROTOCOL: Stream
```
Comment (jcbrinfo) [inline lib/core/protocol.nit]:
```
var protocol: Protocol = …
var stream: protocol.PROTOCOL = protocol.client
```
Hmmm…

>  There are only two hard things in Computer Science: cache invalidation and naming things.
>
>  — Phil Karlton

File: lib/core/protocol.nit [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+	# The client underlying `self`
+	var client: PROTOCOL
+end
+
+# A Write-only protocol, no communication is expected from a client.
+class ProtocolWriter
+	super Protocol
+	super Writer
+	# Which class should a protocol wrap ?
```
Comment (jcbrinfo) [inline lib/core/protocol.nit]: Why repeating the same documentation?

**Comment by R4PaSs** [comment]
I guess this outlived its purpose, closing.

**Pull request closed without merging**